### PR TITLE
Consistently wrap two context managers in parens (in --preview).

### LINKED ISSUE #3572: Parens of parenthesized context managers are removed in some cases, even when targeting 3.9+
Before:

```python
with (
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True
      ),
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
      )
  ):
    pass


with (
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True
      ) as a,
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
      )
  ):
    pass


with (
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True
      ),
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
      ) as b
  ):
    pass


with (
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True
      ),
      mock.patch.object(
          self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
      ),
  ):
    pass
```

After:

```python
with mock.patch.object(self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True), mock.patch.object(
    self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
):
    pass


with mock.patch.object(
    self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True
) as a, mock.patch.object(
    self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
):
    pass


with (
    mock.patch.object(self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True),
    mock.patch.object(
        self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
    ) as b,
):
    pass


with (
    mock.patch.object(self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True),
    mock.patch.object(
        self.xxx_xxxxxx, "xxxxxxxxxx", autospec=True, return_value="xxxx"
    ),
):
    pass
```

[Playground link](https://black.vercel.app/?version=main&state=_Td6WFoAAATm1rRGAgAhARYAAAB0L-Wj4AQqALxdAD2IimZxl1N_WmufDs0Lm7j3vYWBNqJbki6Z3dR6F7CSw5jdn3d7lGX05CVFI8hXkUC-mCeAPAY2ibjlgtMLir9fMIryvcrlWn_huw8eC9Bgu30Rrb76i_ddNbGMYXeC1Ow8DNTmEt8HWMimuIZgqQiR9rniWcdkE3fvUKVarzuNWSFRAKc5meAozYDJD0hr01_4TPwgekwm0EzxyXBCeyPlGPMdaTrcdtomLD284ixkt44xalu9sciwiASdALbZSVGfYZstAAHYAasIAACTrCGUscRn-wIAAAAABFla).

I'll work on a fix.

### PR DESCRIPTION
**Pull request by yilei**
### Description

Fixes #3572

### Checklist - did you ...

- [x] Add an entry in `CHANGES.md` if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation?

### CODE REVIEW

**Comment by yilei** [comment]
Attaching a diff-shades result: [3589.diff.zip](https://github.com/psf/black/files/10855915/3589.diff.zip)

This is from running:

```bash
diff-shades analyze base.json --work-dir projects-cache --force-preview-style -- --target-version=py39
diff-shades analyze new.json --work-dir projects-cache --force-preview-style -- --target-version=py39
diff-shades compare base.json new.json --diff
```

**Comment by yilei** [comment]
Ah, the diff-shades check on CI is broken?

**Comment by ichard26** [comment]
Yup. :upside_down_face:

File: src/black/lines.py [hunk 1]
```diff
@@ -119,6 +120,11 @@ def is_import(self) -> bool:
         """Is this an import line?"""
         return bool(self) and is_import(self.leaves[0])
 
+    @property
+    def is_with_stmt(self) -> bool:
+        """Is this an with_stmt line?"""
```
Comment (JelleZijlstra) [inline src/black/lines.py:125]:
```suggestion
        """Is this a with_stmt line?"""
```

File: CHANGES.md [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 - Add trailing commas to collection literals even if there's a comment after the last
   entry (#3393)
+- `with` statements that contain two context managers will be consistently wrapped in
+  parens (#3589)
```
Comment (JelleZijlstra) [inline CHANGES.md:20]:
```suggestion
  parentheses (#3589)
```
"parens" is slightly too informal for the changelog

File: src/black/nodes.py [hunk 3]
```diff
@@ -789,6 +789,16 @@ def is_import(leaf: Leaf) -> bool:
     )
 
 
+def is_with_stmt(leaf: Leaf) -> bool:
+    """Return True if the given leaf starts a with statement."""
+    return bool(
+        leaf.type == token.NAME
+        and leaf.value == "with"
```
Comment (JelleZijlstra) [inline src/black/nodes.py:796]: Does this also cover `async with`? We should treat both of these the same.

**Review by JelleZijlstra — COMMENTED** [review]
Two nits, and a question about `async with`.

File: src/black/nodes.py (hunk 3 above)
Comment (yilei) [reply src/black/nodes.py:796]: Good catch! Turns out `async with` isn't handled at all in #3489. It has a different code path [here](https://github.com/psf/black/blob/25d886f52c2bbbb58386ac8050f4e67952507bc7/src/black/linegen.py#L289) so the code I'm touching here isn't called at all.

I filed #3591, will address separately.

**Comment by yilei** [comment]
@ichard26 FYI I just sent https://github.com/ichard26/diff-shades/pull/11 to fix diff-shades

**Comment by JelleZijlstra** [comment]
Thanks @yilei! I merged the diff-shades change, let's see if that's enough.

**Comment by ichard26** [comment]
The workflow pulls from the `stable` branch which was admittedly bad design on my part (I'll rename the default branch on ichard26/diff-shades to avoid disrupting this repo). I updated it and triggered a rerun although it seems to be stuck? Let's try this again.

**Comment by yilei** [comment]
Success!

**Comment by yilei** [comment]
Kindly ping @JelleZijlstra?

**Comment by yilei** [comment]
Ah sorry, I'll resolve the merge conflicts.

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**